File: java/exercises/src/test/java/io/github/ajoz/workshop/fp/java/tests/theory/AboveSupplier.java
package io.github.ajoz.workshop.fp.java.tests.theory;

import org.junit.experimental.theories.ParameterSignature;
import org.junit.experimental.theories.ParameterSupplier;
import org.junit.experimental.theories.PotentialAssignment;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public final class AboveSupplier extends ParameterSupplier {
    @Override
    public List<PotentialAssignment> getValueSources(final ParameterSignature sig) throws Throwable {
        final Above above = sig.getAnnotation(Above.class);
        return new Random().ints(
                above.limit(),
                above.inclusive()
                        ? above.value() - 1
                        : above.value(),
                Integer.MAX_VALUE
        )
                .boxed()
                .map(i -> PotentialAssignment.forValue("ints", i))
                .collect(Collectors.toList());
    }
}

File: kotlin/exercises/src/test/kotlin/io/github/ajoz/workshop/fp/kotlin/tests/theory/AboveSupplier.kt
package io.github.ajoz.workshop.fp.kotlin.tests.theory

import org.junit.experimental.theories.ParameterSignature
import org.junit.experimental.theories.ParameterSupplier
import org.junit.experimental.theories.PotentialAssignment
import java.util.Random
import java.util.stream.Collectors

class AboveSupplier : ParameterSupplier() {
    @Throws(Throwable::class)
    override fun getValueSources(sig: ParameterSignature): List<PotentialAssignment> {
        val above: Above = sig.getAnnotation(Above::class.java)
        return Random().ints(
                above.limit.toLong(),
                if (above.inclusive)
                    above.value - 1
                else
                    above.value,
                Integer.MAX_VALUE
        )
                .boxed()
                .map { i -> PotentialAssignment.forValue("ints", i) }
                .collect(Collectors.toList())
    }
}
